cocotb/share/lib/simulator/simulatormodule.cpp

```diff
@@ -41,11 +41,122 @@ static int sim_ending = 0;
 
 #include "simulatormodule.h"
 #include <cocotb_utils.h>     // COCOTB_UNUSED
+#include <type_traits>
+
 
 struct module_state {
     PyObject *error;
 };
 
+
+/* define the extension types as templates */
+namespace {
+    template<typename gpi_hdl>
+    struct gpi_hdl_Object {
+        PyObject_HEAD
+        gpi_hdl hdl;
+
+        // The python type object, in a place that is easy to retrieve in templates
+        static PyTypeObject py_type;
+    };
+
+    /** __repr__ shows the memory address of the internal handle */
+    template<typename gpi_hdl>
+    static PyObject *gpi_hdl_repr(gpi_hdl_Object<gpi_hdl> *self) {
+        auto *type = Py_TYPE(self);
+        return PyUnicode_FromFormat("<%s at %p>", type->tp_name, self->hdl);
+    }
+
+    /** __hash__ returns the pointer itself */
+    template<typename gpi_hdl>
+    static Py_hash_t gpi_hdl_hash(gpi_hdl_Object<gpi_hdl> *self) {
+        auto ret = reinterpret_cast<Py_hash_t>(self->hdl);
+        // hash must never return -1
+        if (ret == (Py_hash_t)-1) {
+            ret = (Py_hash_t)-2;
+        }
+        return ret;
+    }
+
+    /**
+     * Create a new python handle object from a pointer, returning None if the
+     * pointer is NULL.
+     */
+    template<typename gpi_hdl>
+    static PyObject *gpi_hdl_New(gpi_hdl hdl) {
+        if (hdl == NULL) {
+            Py_RETURN_NONE;
+        }
+        auto *obj = PyObject_New(gpi_hdl_Object<gpi_hdl>, &gpi_hdl_Object<gpi_hdl>::py_type);
+        if (obj == NULL) {
+            return NULL;
+        }
+        obj->hdl = hdl;
+        return (PyObject *)obj;
+    }
+
+    /** Comparison checks the types match, and then compares pointers */
+    template<typename gpi_hdl>
+    static PyObject *gpi_hdl_richcompare(PyObject *self, PyObject *other, int op) {
+        if (
+                Py_TYPE(self) != &gpi_hdl_Object<gpi_hdl>::py_type ||
```

Comment:
Looks like clang-format messed up here.

Reply:
Unlikely given I'm not using it! I think some other open source project I contributed too wanted two indent levels for wrapped lines, to distinguish them from nested blocks.

I don't feel strongly, let me know if you want me to unindent this one level.

Reply:
Do you prefer the following?
```C
        if (
            Py_TYPE(self) != &gpi_hdl_Object<gpi_hdl>::py_type ||
            Py_TYPE(other) != &gpi_hdl_Object<gpi_hdl>::py_type
        ) {
            Py_RETURN_NOTIMPLEMENTED;
        }
```

Reply:
Went ahead and used that approach, since it was quicker than awaiting a response.

Reply:
I'd go for 
```
        if (Py_TYPE(self) != &gpi_hdl_Object<gpi_hdl>::py_type ||
            Py_TYPE(other) != &gpi_hdl_Object<gpi_hdl>::py_type
        ) {
            Py_RETURN_NOTIMPLEMENTED;
        }
```

but either way is fine.

Reply:
Not a fan of that because it's quite hard to distinguish at a glance from the nonsensical but legal:
```C++
        if (Py_TYPE(self) != &gpi_hdl_Object<gpi_hdl>::py_type) {
            Py_TYPE(other) != &gpi_hdl_Object<gpi_hdl>::py_type;
            Py_RETURN_NOTIMPLEMENTED;
        }
```